src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -46832,6 +46832,226 @@ class CL1
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x1").WithLocation(11, 15));
         }
 
+        /// <remarks>Ported from <see cref="FlowTests.IfConditionalConstant" />.</remarks>
+        [Theory]
+        [InlineData("true", "false", "object?")]
+        [InlineData("true", "false", "object")]
+        [InlineData("FIELD_TRUE", "FIELD_FALSE", "object?")]
+        [InlineData("LOCAL_TRUE", "LOCAL_FALSE", "object?")]
+        [InlineData("true || false", "true && false", "object?")]
+        [InlineData("!false", "!true", "object?")]
+        public void IfConditionalConstant(string @true, string @false, string parameterType)
+        {
+            var source = @"
+#pragma warning disable 219 // The variable is assigned but its value is never used
+
+class C
+{
+    const bool FIELD_TRUE = true;
+    const bool FIELD_FALSE = false;
+
+    static void M(bool b, " + parameterType + @" x, object y)
+    {
+        const bool LOCAL_TRUE = true;
+        const bool LOCAL_FALSE = false;
+
+        {
+            if (b ? x != null : " + @false + @")
+                y = x;
+        }
+
+        {
+            if (b ? x != null : " + @true + @")
+                y = x; // 1
+        }
+
+        {
+            if (b ? " + @false + @" : x != null)
+                y = x;
+        }
+
+        {
+            if (b ? " + @true + @" : x != null)
+                y = x; // 2
+        }
+
+        {
+            if (!(b ? " + @false + @" : x == null))
+                y = x; // 3
```

Comment:
>// 3 [](http://example.com/codeflow?start=23&length=4)

I'm confused how we get here when the type of x is `object`: I think it's because the previous `if` set the state to `MaybeNull` for all cases. I think we need to add another parameter to assign to `x` at the beginning of each block to reset it's state to the expected default.

Reply:
Looking into this I realize I'm not quite satisfied with the coding pattern used in this test, so I rewrote the test while ensuring that the scenarios you've called out are covered.